File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/fs/vgmrips/Identifier.java
/**
 * @file
 * @brief
 * @author vitamin.caig@gmail.com
 */

package app.zxtune.fs.vgmrips;

import android.net.Uri;

import androidx.annotation.Nullable;

import java.util.List;

import app.zxtune.TimeStamp;

/**
 * Paths:
 * <p>
 * 1) vgmrips:/
 * 2) vgmrips:/Company
 * 3) vgmrips:/Company/${company_name}?group=${company_id}
 * 4) vgmrips:/Composer
 * 5) vgmrips:/Composer/${composer_name}?group=${composer_id}
 * 6) vgmrips:/Chip
 * 7) vgmrips:/Chip/${chip_name}?group=${chip_id}
 * 8) vgmrips:/System
 * 9) vgmrips:/System/${system_name}?group=${system_id}
 * 10) vgmrips://Random
 * <p>
 * X) ${url}&pack=${pack_id}
 */

public class Identifier {

  private static final String SCHEME = "vgmrips";

  private static final int POS_CATEGORY = 0;
  private static final int POS_GROUP_NAME = 1;
  private static final int POS_PACK_NAME = 2;
  private static final int POS_RANDOM_PACK_NAME = 1;

  private static final String PARAM_GROUP = "group";
  private static final String PARAM_PACK = "pack";
  private static final String PARAM_TRACK = "track";

  public static final String CATEGORY_COMPANY = "Company";
  public static final String CATEGORY_COMPOSER = "Composer";
  public static final String CATEGORY_CHIP = "Chip";
  public static final String CATEGORY_SYSTEM = "System";
  public static final String CATEGORY_RANDOM = "Random";

  // Root
  public static Uri.Builder forRoot() {
    return new Uri.Builder().scheme(SCHEME);
  }

  public static boolean isFromRoot(Uri uri) {
    return SCHEME.equals(uri.getScheme());
  }

  // Categories
  public static Uri.Builder forCategory(String category) {
    return forRoot().appendPath(category);
  }

  @Nullable
  public static String findCategory(List<String> path) {
    return path.size() > POS_CATEGORY ? path.get(POS_CATEGORY) : null;
  }

  public static Uri.Builder forGroup(Uri.Builder parent, Group obj) {
    return parent
        .appendPath(obj.getTitle())
        .appendQueryParameter(PARAM_GROUP, obj.getId());
  }

  @Nullable
  public static Group findGroup(Uri uri, List<String> path) {
    if (path.size() > POS_GROUP_NAME) {
      final String id = uri.getQueryParameter(PARAM_GROUP);
      if (id != null) {
        final String name = path.get(POS_GROUP_NAME);
        return new Group(id, name, 0/*fake*/);
      }
    }
    return null;
  }

  public static Uri.Builder forPack(Uri.Builder parent, Pack pack) {
    return parent.appendPath(pack.getTitle()).appendQueryParameter(PARAM_PACK, pack.getId());
  }

  @Nullable
  public static Pack findPack(Uri uri, List<String> path) {
    if (path.size() > POS_PACK_NAME) {
      final String id = uri.getQueryParameter(PARAM_PACK);
      if (id != null) {
        final String name = path.get(POS_PACK_NAME);
        return new Pack(id, name);
      }
    }
    return null;
  }

  @Nullable
  public static Pack findRandomPack(Uri uri, List<String> path) {
    if (path.size() > POS_RANDOM_PACK_NAME) {
      final String id = uri.getQueryParameter(PARAM_PACK);
      if (id != null) {
        final String name = path.get(POS_RANDOM_PACK_NAME);
        return new Pack(id, name);
      }
    }
    return null;
  }

  public static Uri.Builder forTrack(Uri.Builder parent, Track track) {
    return parent.appendPath(track.getTitle()).appendQueryParameter(PARAM_TRACK, track.getLocation());
  }

  @Nullable
  public static Track findTrack(Uri uri, List<String> path) {
    final int elements = path.size();
    if (elements >= 2) {
      final String id = uri.getQueryParameter(PARAM_TRACK);
      if (id != null) {
        final String name = path.get(elements - 1);
        return new Track(0/*fake*/, name, TimeStamp.EMPTY/*fake*/, id);
      }
    }
    return null;
  }
}

File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/fs/vgmrips/Identifier.kt
/**
 * @file
 * @brief
 * @author vitamin.caig@gmail.com
 */
package app.zxtune.fs.vgmrips

import android.net.Uri
import androidx.annotation.VisibleForTesting
import app.zxtune.utils.ifNotNulls

/**
 * Paths:
 *
 *
 * 1) vgmrips:/
 * 2) vgmrips:/Company
 * 3) vgmrips:/Company/${company_name}?group=${company_id}
 * 4) vgmrips:/Composer
 * 5) vgmrips:/Composer/${composer_name}?group=${composer_id}
 * 6) vgmrips:/Chip
 * 7) vgmrips:/Chip/${chip_name}?group=${chip_id}
 * 8) vgmrips:/System
 * 9) vgmrips:/System/${system_name}?group=${system_id}
 * 10) vgmrips:/Random
 *
 *
 * X) ${url}&pack=${pack.id}&location=${pack.archive}
 *
 * Track identifiers (for compatibility):
 * .../${pack.title}/${track.title}?pack=${pack.id}&track=${track.location}
 *
 * Images:
 * vgmrips:/Image?location=${path}
 *
 * Check query params at the end - UOE if uri is not hierarchical (root)
 */
object Identifier {
    private const val SCHEME = "vgmrips"
    private const val POS_CATEGORY = 0
    private const val POS_GROUP_NAME = 1
    private const val POS_PACK_NAME = 2
    private const val PARAM_GROUP = "group"
    private const val PARAM_PACK = "pack"
    private const val PARAM_TRACK = "track"
    private const val PARAM_LOCATION = "location"
    const val CATEGORY_COMPANY = "Company"
    const val CATEGORY_COMPOSER = "Composer"
    const val CATEGORY_CHIP = "Chip"
    const val CATEGORY_SYSTEM = "System"
    const val CATEGORY_RANDOM = "Random"
    const val CATEGORY_IMAGE = "Image"

    // Root
    fun forRoot(): Uri.Builder = Uri.Builder().scheme(SCHEME)

    fun isFromRoot(uri: Uri) = SCHEME == uri.scheme

    // Categories
    fun forCategory(category: String): Uri.Builder = forRoot().appendPath(category)

    fun findCategory(path: List<String>) = path.getOrNull(POS_CATEGORY)

    fun forGroup(parent: Uri.Builder, obj: Group) = parent.apply {
        appendPath(obj.title)
        appendQueryParameter(PARAM_GROUP, obj.id.value)
    }

    fun findGroup(uri: Uri, path: List<String>) = path.getOrNull(POS_GROUP_NAME)?.let { name ->
        uri.getQueryParameter(PARAM_GROUP)?.let { id ->
            Group(Group.Id(id), name, 0 /*fake*/)
        }
    }

    fun forPack(parent: Uri.Builder, pack: Pack) = parent.apply {
        appendPath(pack.title)
        appendQueryParameter(PARAM_PACK, pack.id.value)
        appendQueryParameter(PARAM_LOCATION, pack.archive.value)
    }

    @VisibleForTesting
    fun findPackId(uri: Uri) =
        uri.takeIf { it.isHierarchical }?.getQueryParameter(PARAM_PACK)?.let {
            Pack.Id(it)
        }

    fun findPack(uri: Uri, path: List<String>) = path.getOrNull(POS_PACK_NAME)?.let { name ->
        ifNotNulls(findPackId(uri), uri.getQueryParameter(PARAM_LOCATION)) { id, archive ->
            Pack(id, name, FilePath(archive))
        }
    }

    fun forRandomTrack(pack: Pack.Id, track: FilePath): Uri =
        forCategory(CATEGORY_RANDOM).appendQueryParameter(PARAM_PACK, pack.value)
            .appendQueryParameter(PARAM_TRACK, track.value).build()

    fun findTrack(uri: Uri) = findPackId(uri)?.let { id ->
        uri.getQueryParameter(PARAM_TRACK)?.let { track ->
            Pair(id, FilePath(track))
        }
    }

    fun forImage(path: FilePath): Uri =
        forCategory(CATEGORY_IMAGE).appendQueryParameter(PARAM_LOCATION, path.value).build()

    fun findImage(uri: Uri) = uri.getQueryParameter(PARAM_LOCATION)?.let {
        require(uri.lastPathSegment == CATEGORY_IMAGE)
        FilePath(it)
    }
}
